### PR TITLE
Fix #2669 and #2672

### LINKED ISSUE #2669: SameSelection should have an option for indices
## Expected behavior ##

`byres` and `same residue as` selects the whole residue for each atom in the subsequent selection, by the canonical and unique resindex. Likewise `same segment as` uses `segindex`. 

## Actual behavior ##

`byres` and `same residue as` use the user-supplied, potentially repeated `resid` values. `same segment as` uses `segid`. This is redundant as you can already use `same resid as` or `same segid as` for the same behaviour. This may result in unexpected behaviour with altLocs also. 

## Code to reproduce the behavior ##

Leads to problems like in this [mailing list issue](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/mdnalysis-discussion/oY7pdg8375Y/4mtLSkJBBAAJ) where selecting "byres resname TIP3" returns various amino acids and ions as well.

```python
>>> import MDAnalysis as mda
>>> u = mda.Universe('a.psf', 'a.dcd')
>>> ag = u.select_atoms('byres resname TIP3')
>>> print(set(ag.resnames))
{'GLN', 'LEU', 'TRP', 'THR', 'LYS', 'GLU', 'POPE', 'ASP', 'TYR', 'VAL', 'ARG', 'PHE', 'MET', 'CYS', 'ILE', 'PRO', 'TIP3', 'ALA', 'SOD', 'CLA', 'SER', 'GLY', 'ASN', 'HSD'}
```

## Current version of MDAnalysis ##

- Which version are you using? (run `python -c "import MDAnalysis as mda; print(mda.__version__)"`) 0.20.1
- Which version of Python (`python -V`)? 3.7.3
- Which operating system? MacOS

### LINKED ISSUE #2672: byres selection fails with insertion code
## Expected behavior ##

Correct `byres` selection for PDB files.

## Actual behavior ##

Using `byres` with PDB files containing insertion codes leads to erroneous selections, where all residues with the same number but different insertion codes are selected.

This is linked to #2308.

## Code to reproduce the behavior ##

``` python
In [1]: import MDAnalysis as mda

In [2]: u = mda.Universe("p.pdb")

In [3]: ui = mda.Universe("p_icode.pdb")

In [4]: u.select_atoms("byres(around 2.0 resnum 13)")
Out[4]: <AtomGroup with 21 atoms>

In [5]: ui.select_atoms("byres(around 2.0 resnum 13)")
Out[5]: <AtomGroup with 12 atoms>
```

with `p_icode.pd`

```
ATOM    217  N   GLU L  13      -5.997  11.588  22.006  1.00 22.34           N
ATOM    218  H   GLU L  13      -7.033  11.673  22.032  1.00  0.00           H
ATOM    219  CA  GLU L  13      -5.303  10.472  22.709  1.00 24.14           C
ATOM    220  HA  GLU L  13      -4.258  10.727  22.888  1.00  0.00           H
ATOM    221  C   GLU L  13      -5.380   9.199  21.913  1.00 23.81           C
ATOM    222  O   GLU L  13      -6.470   8.998  21.302  1.00 25.39           O
ATOM    223  CB  GLU L  13      -6.150  10.222  23.967  1.00 26.06           C
ATOM    224  HB1 GLU L  13      -7.200  10.170  23.680  1.00  0.00           H
ATOM    225  HB2 GLU L  13      -5.849   9.274  24.412  1.00  0.00           H
ATOM    226  CG  GLU L  13      -5.978  11.320  24.994  1.00 36.69           C
ATOM    227  HG1 GLU L  13      -4.919  11.576  25.032  1.00  0.00           H
ATOM    228  HG2 GLU L  13      -6.548  12.185  24.654  1.00  0.00           H
ATOM    229  CD  GLU L  13      -6.421  11.003  26.391  1.00 42.27           C
ATOM    230  OE1 GLU L  13      -6.995   9.892  26.528  1.00 46.63           O
ATOM    231  OE2 GLU L  13      -6.208  11.780  27.305  1.00 46.31           O
ATOM    232  N   ASP L  14      -4.407   8.330  21.972  1.00 22.45           N
ATOM    233  H   ASP L  14      -3.528   8.586  22.465  1.00  0.00           H
ATOM    234  CA  ASP L  14      -4.526   6.972  21.347  1.00 21.73           C
ATOM    235  HA  ASP L  14      -5.031   6.955  20.381  1.00  0.00           H
ATOM    236  C   ASP L  14      -5.408   6.153  22.306  1.00 20.49           C
ATOM    237  O   ASP L  14      -5.818   6.528  23.416  1.00 19.90           O
ATOM    238  CB  ASP L  14      -3.184   6.429  20.919  1.00 20.07           C
ATOM    239  HB1 ASP L  14      -3.360   5.540  20.314  1.00  0.00           H
ATOM    240  HB2 ASP L  14      -2.694   7.189  20.311  1.00  0.00           H
ATOM    241  CG  ASP L  14      -2.229   6.045  22.052  1.00 17.98           C
ATOM    242  OD1 ASP L  14      -2.698   5.227  22.871  1.00 21.20           O
ATOM    243  OD2 ASP L  14      -1.046   6.434  21.950  1.00 20.58           O
ATOM    244  N   LYS L  14A     -5.864   5.031  21.805  1.00 21.06           N
ATOM    245  H   LYS L  14A     -5.496   4.733  20.879  1.00  0.00           H
ATOM    246  CA  LYS L  14A     -6.839   4.178  22.443  1.00 21.93           C
ATOM    247  HA  LYS L  14A     -7.640   4.846  22.760  1.00  0.00           H
ATOM    248  C   LYS L  14A     -6.366   3.509  23.720  1.00 23.19           C
ATOM    249  O   LYS L  14A     -7.278   3.215  24.547  1.00 24.43           O
ATOM    250  CB  LYS L  14A     -7.402   3.138  21.440  1.00 26.01           C
ATOM    251  HB1 LYS L  14A     -7.879   3.658  20.609  1.00  0.00           H
ATOM    252  HB2 LYS L  14A     -6.587   2.520  21.063  1.00  0.00           H
```
and `p.pdb`

```
ATOM    217  N   GLU L  13      -5.997  11.588  22.006  1.00 22.34           N
ATOM    218  H   GLU L  13      -7.033  11.673  22.032  1.00  0.00           H
ATOM    219  CA  GLU L  13      -5.303  10.472  22.709  1.00 24.14           C
ATOM    220  HA  GLU L  13      -4.258  10.727  22.888  1.00  0.00           H
ATOM    221  C   GLU L  13      -5.380   9.199  21.913  1.00 23.81           C
ATOM    222  O   GLU L  13      -6.470   8.998  21.302  1.00 25.39           O
ATOM    223  CB  GLU L  13      -6.150  10.222  23.967  1.00 26.06           C
ATOM    224  HB1 GLU L  13      -7.200  10.170  23.680  1.00  0.00           H
ATOM    225  HB2 GLU L  13      -5.849   9.274  24.412  1.00  0.00           H
ATOM    226  CG  GLU L  13      -5.978  11.320  24.994  1.00 36.69           C
ATOM    227  HG1 GLU L  13      -4.919  11.576  25.032  1.00  0.00           H
ATOM    228  HG2 GLU L  13      -6.548  12.185  24.654  1.00  0.00           H
ATOM    229  CD  GLU L  13      -6.421  11.003  26.391  1.00 42.27           C
ATOM    230  OE1 GLU L  13      -6.995   9.892  26.528  1.00 46.63           O
ATOM    231  OE2 GLU L  13      -6.208  11.780  27.305  1.00 46.31           O
ATOM    232  N   ASP L  14      -4.407   8.330  21.972  1.00 22.45           N
ATOM    233  H   ASP L  14      -3.528   8.586  22.465  1.00  0.00           H
ATOM    234  CA  ASP L  14      -4.526   6.972  21.347  1.00 21.73           C
ATOM    235  HA  ASP L  14      -5.031   6.955  20.381  1.00  0.00           H
ATOM    236  C   ASP L  14      -5.408   6.153  22.306  1.00 20.49           C
ATOM    237  O   ASP L  14      -5.818   6.528  23.416  1.00 19.90           O
ATOM    238  CB  ASP L  14      -3.184   6.429  20.919  1.00 20.07           C
ATOM    239  HB1 ASP L  14      -3.360   5.540  20.314  1.00  0.00           H
ATOM    240  HB2 ASP L  14      -2.694   7.189  20.311  1.00  0.00           H
ATOM    241  CG  ASP L  14      -2.229   6.045  22.052  1.00 17.98           C
ATOM    242  OD1 ASP L  14      -2.698   5.227  22.871  1.00 21.20           O
ATOM    243  OD2 ASP L  14      -1.046   6.434  21.950  1.00 20.58           O
ATOM    244  N   LYS L  15     -5.864   5.031  21.805  1.00 21.06           N
ATOM    245  H   LYS L  15     -5.496   4.733  20.879  1.00  0.00           H
ATOM    246  CA  LYS L  15     -6.839   4.178  22.443  1.00 21.93           C
ATOM    247  HA  LYS L  15     -7.640   4.846  22.760  1.00  0.00           H
ATOM    248  C   LYS L  15     -6.366   3.509  23.720  1.00 23.19           C
ATOM    249  O   LYS L  15     -7.278   3.215  24.547  1.00 24.43           O
ATOM    250  CB  LYS L  15     -7.402   3.138  21.440  1.00 26.01           C
ATOM    251  HB1 LYS L  15     -7.879   3.658  20.609  1.00  0.00           H
ATOM    252  HB2 LYS L  15     -6.587   2.520  21.063  1.00  0.00           H
```

## Current version of MDAnalysis ##

- Which version are you using? 0.20.1
- Which version of Python (`python -V`)? 3.7.6
- Which operating system? macOS

### PR DESCRIPTION
**Pull request by RMeli**
Fixes #2669 and fixes #2672

Changes made in this Pull Request:
 - Use `resindices` of `resids` for `same residue` selection
 - Use `segindices` of `segids` for `same segment` selection
 - Use `resindices` for `byres` selection

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

**Comment by RMeli** [comment]
Restarted Travis-CI build #2240.8 which timed out (see #2671).

File: testsuite/MDAnalysisTests/core/test_atomselections.py [hunk 1]
```diff
@@ -1101,6 +1101,31 @@ def test_missing_icodes_range_VE(self, u):
...
+        # See Issues #2308 for a discussion
+        ("same resid as", 72),
+        # Selection using resindices
+        # For PDBs: residues with different insertion codes have different resindices
+        ("byres", 11)
+    ]
+)
+def test_similarity_selection_icodes(selection, n_atoms):
+
+    u = mda.Universe(PDB_icodes)
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/core/test_atomselections.py:1121]: make this a fixture

Comment (RMeli) [reply testsuite/MDAnalysisTests/core/test_atomselections.py]: @orbeckst Done (if what I did is what you had in mind).

**Comment by orbeckst** [comment]
Many thanks @RMeli !

**Pull request merged**